### PR TITLE
Clean up Hash Functions

### PR DESCRIPTION
**Pull request by reduz**
Clean up and do fixes to hash functions and newly introduced murmur3 hashes in #61934
* Clean up usage of murmur3
* Fixed usages of binary murmur3 on floats (this is invalid because -0/0 as well as inf must be taken into consideration).
* Changed DJB2 to use xor (which seems to be better)

### CODE REVIEW

File: core/templates/hashfuncs.h [hunk 1]
```diff
@@ -241,31 +303,69 @@ struct HashMapHasherDefault {
...
+	static _FORCE_INLINE_ uint32_t hash(const float p_float) { return hash_murmur3_one_float(p_float); }
+	static _FORCE_INLINE_ uint32_t hash(const double p_double) { return hash_murmur3_one_double(p_double); }
+	static _FORCE_INLINE_ uint32_t hash(const uint32_t p_int) { return hash_fmix32(p_int); }
+	static _FORCE_INLINE_ uint32_t hash(const int32_t p_int) { return hash_fmix32(p_int); }
+	static _FORCE_INLINE_ uint32_t hash(const uint16_t p_int) { return hash_fmix32(p_int); }
+	static _FORCE_INLINE_ uint32_t hash(const int16_t p_int) { return hash_fmix32(p_int); }
+	static _FORCE_INLINE_ uint32_t hash(const uint8_t p_int) { return hash_fmix32(p_int); }
+	static _FORCE_INLINE_ uint32_t hash(const int8_t p_int) { return hash_fmix32(p_int); }
+	static _FORCE_INLINE_ uint32_t hash(const Vector2i &p_vec) {
+		uint32_t h = hash_murmur3_one_real(p_vec.x);
```
Comment (akien-mga) [inline core/templates/hashfuncs.h:325]: Copy paste mistake?

Comment (reduz) [reply core/templates/hashfuncs.h:325]: ah yes thanks

**Review by Geometror — COMMENTED** [review]
Looks good to me:)
I wonder if we could use murmur3 for string hashing too? Based on [this ](https://softwareengineering.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed) actually pretty good stackexchange answer it should perform better than DBJ2, other benchmarks show similar results (though I am not sure how it behaves with short strings). During my experiments with different hash functions I tried to do that, but the engine crashed since it somehow broke ClassDB.

**Comment by lmurray** [comment]
> During my experiments with different hash functions I tried to do that, but the engine crashed since it somehow broke ClassDB.

StringName makes an assumption that all encodings of a string will result with the same hash, e.g. with a `char *` vs a `char32_t *`. This is only true with the current DJB2 algorithm with ASCII characters which all StringName strings currently are.

It's also worth noting that Murmur3A is slower than the DJB2 algorithm for strings under ~16 characters or so. I've been doing my own tests with hashes and have yet to find a good solution that performs better than the current version. I'm currently of the impression that just using a higher-quality minimal algorithm such as FNV-1a everywhere will end up being more performant than trying to use a heavier algorithm in certain circumstances due to being easier to inline.

File: core/variant/variant_call.cpp [hunk 2]
```diff
@@ -1170,19 +1170,19 @@ uint32_t Variant::get_builtin_method_hash(Variant::Type p_type, const StringName
...
+	hash = hash_murmur3_one_32(method->is_vararg, hash);
+	hash = hash_murmur3_one_32(method->has_return_type, hash);
 	if (method->has_return_type) {
-		hash = hash_djb2_one_32(method->return_type, hash);
+		hash = hash_murmur3_one_32(method->return_type, hash);
 	}
-	hash = hash_djb2_one_32(method->argument_count, hash);
+	hash = hash_murmur3_one_32(method->argument_count, hash);
 	for (int i = 0; i < method->argument_count; i++) {
 		hash = method->get_argument_type(i);
```
Comment (lmurray) [inline core/variant/variant_call.cpp:1182]: The hash is mistakenly overwritten here.

**Comment by reduz** [comment]
@lmurray @Geometror To be honest I think for strings DJB2 does the job. The reason I moved most stuff to murmur3 is to improve the distribution in robin hood hash tables (which are super fast, but require in exchange much better hashing than what we had, specially when using sequential elements, which DJB2 kind of sucks for). But for strings it should be fine, I don't see distribution being a problem in that case, so I doubt it makes much sense.

**Review by Geometror — APPROVED** [review]
Should be good to go, we can always change these if we find something which has a noticeable performance benefit over the combination of murmur3 and dbj2.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Zireael07** [comment]
Side note: are the new hash() functions exposed to the scripting? I need to hash a Vector3...

**Comment by Calinou** [comment]
> Side note: are the new hash() functions exposed to the scripting? I need to hash a Vector3...

It's not exposed for Vector2/3(i):

image

**Pull request merged**